### PR TITLE
Fix listeners leak when using with componentDidCatch

### PR DESCRIPTION
**Pull request by SamyPesse**
⚠️ *This PR is not ready to be merged, neither tested. I'm opening it now to gather feedback and suggestions of solutions that I may have missed.*

I investigated a bug in my application coming from the use of `unstated` and `componentDidCatch`.

At some point `container.setState` was returning a never-fulfilled promise. React was logging:

```
warning.js:33 Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in Subscribe
    in Fetch
    in FetchBoundary (created by ViewTable)
    in ViewTable (created by SecondaryView)
    in SecondaryView
```

When an error is thrown during the render phase, the [`componentWillUnmount`](https://github.com/jamiebuilds/unstated/blob/master/src/unstated.js#L81) is never called, and the component doesn't unsubscribe from the container.

When calling `setState`, since the component for the listener is unmounted, [`onUpdate` returns a never-fulfilled promise](https://github.com/jamiebuilds/unstated/blob/master/src/unstated.js#L92).

**Potential solution:** I'm experimenting moving the subscribe calls to a safe place `componentDidMount` (which happens during the commit phase, see [this comment](https://github.com/gaearon/react-side-effect/issues/40#issuecomment-367129253) that explains it).

Any way, I'd love to get your feedback on what could be the best approach to fix this 🙌

---

TLDR: When using `unstated` with `componentDidCatch`, it leaks listeners, and prevent `setState` from being fulfilled.

### CODE REVIEW

File: src/unstated.js [hunk 1]
```diff
@@ -64,24 +64,45 @@ export type SubscribeProps<Containers: ContainersType> = {
...
+  ) => Node
+};
+type SubscribeUpdaterState = {};
 
 const DUMMY_STATE = {};
 
-export class Subscribe<Containers: ContainersType> extends React.Component<
-  SubscribeProps<Containers>,
-  SubscribeState
+class SubscribeUpdater<Containers: ContainersType> extends React.Component<
```
Comment (SamyPesse) [inline src/unstated.js]: I'm not a big fan of splitting the `Subscribe` component in two components, but because of the context API, it looked like the cleaner method to handle the lifecycle of the instances (subscribe to new instances, unsubscribe from previous one when it render, etc).

**Comment by SamyPesse** [comment]
This solution is not working, because it changes the order of listeners (child component subscribe before the parent does).

I'm gonna investigate other solutions...
